### PR TITLE
Connection: Add site creation date argument

### PR DESCRIPTION
**Pull request by tyxla**
As part of our Jetpack Improved Onboarding work, we'd like to send an additional argument with the rest of the connection URL arguments. That argument would contain the site creation date, which is basically the earliest of either: 

* The earliest admin user registration date.
* The earliest post date.

Companion diff: D25012-code.

#### Changes proposed in this Pull Request:

* Introduce a `Jetpack::get_assumed_site_creation_date()` method.
* Add a new `site_created` parameter to the connection URLs.
* Add an exception for multisite subsites, where we use the multisite creation date instead.

#### Testing instructions:

* Follow the testing instructions in D25012-code.
* Checkout this branch.
* Verify tests pass by running either:

```
yarn docker:phpunit --filter test_get_assumed_site_creation_date
```

or:

```
phpunit --filter test_get_assumed_site_creation_date
```

depending on the environment you're developing Jetpack in.

#### Proposed changelog entry for your changes:

* No changelog entry needed.

### CODE REVIEW

**Comment by kraftbj** [comment]
Would it make sense to query the oldest post object and use the oldest between that and the oldest user?

1. I make a site in 2015 at my company. I was the only person messing with the site.
2. New person is hired in 2018 to help out.
3. I quit in 2019. My user is deleted, authorship changed to user 2.
4. The oldest user is now 2018, even though the site has been around since 2015.

File: class.jetpack.php [hunk 1]
```diff
@@ -4844,6 +4845,24 @@ function build_connect_url( $raw = false, $redirect = false, $from = false, $reg
...
 	}
 
+	/**
+	 * Get the site age, based on the earliest user registration date.
+	 *
+	 * @since 7.1.0
+	 *
+	 * @return string Site creation date and time.
+	 */
+	public static function get_site_creation_date() {
```
Comment (dereksmart) [inline class.jetpack.php]: This has enough chance to be a misnomer (related to @kraftbj's comment) that I would recommend we rename it explicitly to `get_earliest_user_registration_date()`

**Review by AnnaMag — COMMENTED** [review]
Curious: could we pass a classifier (new vs established site) instead of the actual date— `site_age_category` vs `site_creation_date`? Do we need such detail info to be passed around?

**Comment by tyxla** [comment]
Thanks everybody for the feedback!

> Would it make sense to query the oldest post object and use the oldest between that and the oldest user?

Thanks @kraftbj, that scenario makes sense to me. I've addressed that in https://github.com/Automattic/jetpack/pull/11450/commits/0a4a4cbcbc066ac150eef79c39d2732a20f05ec3 - would appreciate feedback on the approach.

> This has enough chance to be a misnomer (related to @kraftbj's comment) that I would recommend we rename it explicitly to get_earliest_user_registration_date()

Thanks @dereksmart - this suggestion made sense, but maybe not after I added the earliest post date in the mix. Just to avoid any confusion you described, I've renamed it to `Jetpack::get_assumed_site_creation_date()`. How does that sound? Do you have any better recommendations?

> Curious: could we pass a classifier (new vs established site) instead of the actual date— site_age_category vs site_creation_date? Do we need such detail info to be passed around?

I think a classifier wouldn't be helpful, because a new site will no longer be a new site after a month, so that classifier could become invalid in some time. Instead, if we store a date, we can do whatever we want with it on the dotcom side (including, automatically maintain an up-to-date classifier as you mentioned with an automated job). I think that the classifier idea is a good idea, but it doesn't make sense to be done here - I'd see it done on the .com side, where we still compute it from the site creation date that we pass here. 

Now that this PR is punted for 7.2, I'd like to add some tests and further manually test it (which is why I marked it as "in-progress". Feedback on the approach is very welcome though!

**Comment by dereksmart** [comment]
`get_assumed_site_creation_date` sounds better. 

It may end up growing a bit as we add more "guess factors" to the mix, but if we need to we can split out the method later.

File: class.jetpack.php [hunk 2]
```diff
@@ -7474,3 +7513,5 @@ public static function show_backups_ui() {
 		return Jetpack::is_plugin_active( 'vaultpress/vaultpress.php' ) || apply_filters( 'jetpack_show_backups', true );
 	}
 }
+
+Jetpack::get_assumed_site_creation_date();
```
Comment (kraftbj) [inline class.jetpack.php]: Is this just a bit of leftover code?

File: class.jetpack-network.php [hunk 3]
```diff
@@ -476,6 +476,7 @@ public function do_subsiteregister( $site_id = null ) {
 				'state'                 => $user_id,
 				'_ui'                   => $tracks_identity['_ui'],
 				'_ut'                   => $tracks_identity['_ut'],
+				'site_created'          => Jetpack::get_assumed_site_creation_date(),
```
[the remark this answers is not in the document]
Comment (tyxla) [reply class.jetpack-network.php]: Nice 👍 I like that, implemented in https://github.com/Automattic/jetpack/pull/11450/commits/51989c4f12425f00348c954e848064839be213f2

File: class.jetpack.php (hunk 2 above)
Comment (tyxla) [reply class.jetpack.php]: Yeah, cleaned up in https://github.com/Automattic/jetpack/pull/11450/commits/8af97d252b0f3dd0db2c60120c0fcbf6af54a82d.

**Comment by tyxla** [comment]
Thanks everyone for the feedback! This is ready for another review.

**Review by AnnaMag — APPROVED** [review]
LGTM

**Review by jeherve — APPROVED** [review]
This works well for me. 🚢

**Comment by tyxla** [comment]
Thank you, folks!

**Pull request merged**